File: python/open_guardrail/guards/idempotent_response.py
"""Detect potentially non-idempotent responses."""
from __future__ import annotations

import re
import time

from open_guardrail.core import GuardResult

_TIMESTAMP: list[re.Pattern[str]] = [
    re.compile(
        r"\b\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}"
    ),
    re.compile(
        r"\b\d{1,2}:\d{2}(?::\d{2})?"
        r"\s*(?:AM|PM)\b",
        re.IGNORECASE,
    ),
    re.compile(
        r"\b(?:today|now)\s+(?:is|at)\s+",
        re.IGNORECASE,
    ),
]

_TEMPORAL: list[re.Pattern[str]] = [
    re.compile(r"\bjust\s+now\b", re.IGNORECASE),
    re.compile(r"\bright\s+now\b", re.IGNORECASE),
    re.compile(
        r"\bat\s+this\s+(?:very\s+)?moment\b",
        re.IGNORECASE,
    ),
    re.compile(
        r"\bcurrently\s+(?:it\s+is|the\s+time)\b",
        re.IGNORECASE,
    ),
    re.compile(
        r"\bas\s+of\s+(?:right\s+)?now\b",
        re.IGNORECASE,
    ),
    re.compile(
        r"\bat\s+the\s+time\s+of"
        r"\s+(?:writing|this\s+response)\b",
        re.IGNORECASE,
    ),
]

_UUID = re.compile(
    r"\b[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}"
    r"-[0-9a-f]{4}-[0-9a-f]{12}\b",
    re.IGNORECASE,
)

_RANDOM_NUM = re.compile(
    r"\b(?:random|generated)\s*(?::?\s*)\d+\b",
    re.IGNORECASE,
)


class _IdempotentResponse:
    def __init__(self, *, action: str = "warn") -> None:
        self.name = "idempotent-response"
        self.action = action

    def check(
        self, text: str, stage: str = "output"
    ) -> GuardResult:
        start = time.perf_counter()
        issues: list[str] = []

        if any(p.search(text) for p in _TIMESTAMP):
            issues.append("timestamp_present")
        if any(p.search(text) for p in _TEMPORAL):
            issues.append("temporal_reference")
        if _UUID.search(text):
            issues.append("uuid_present")
        if _RANDOM_NUM.search(text):
            issues.append("random_number")

        triggered = len(issues) > 0
        score = (
            min(len(issues) / 3, 1.0)
            if triggered
            else 0.0
        )
        elapsed = (time.perf_counter() - start) * 1000

        return GuardResult(
            guard_name="idempotent-response",
            passed=not triggered,
            action=self.action if triggered else "allow",
            score=score,
            message=(
                "Non-idempotent response detected"
                if triggered
                else None
            ),
            latency_ms=round(elapsed, 2),
            details=(
                {"issues": issues} if triggered else None
            ),
        )


def idempotent_response(
    *, action: str = "warn"
) -> _IdempotentResponse:
    return _IdempotentResponse(action=action)

File: packages/guards/src/idempotent-response.ts
import type { Guard, GuardResult, GuardContext } from 'open-guardrail-core';

interface IdempotentResponseOptions {
  action: 'block' | 'warn';
}

const TIMESTAMP_PATTERNS: RegExp[] = [
  /\b\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}/,
  /\b\d{1,2}:\d{2}(?::\d{2})?\s*(?:AM|PM)\b/i,
  /\b(?:today|now)\s+(?:is|at)\s+/i,
];

const TEMPORAL_PATTERNS: RegExp[] = [
  /\bjust\s+now\b/i,
  /\bright\s+now\b/i,
  /\bat\s+this\s+(?:very\s+)?moment\b/i,
  /\bcurrently\s+(?:it\s+is|the\s+time)\b/i,
  /\bas\s+of\s+(?:right\s+)?now\b/i,
  /\bat\s+the\s+time\s+of\s+(?:writing|this\s+response)\b/i,
];

const UUID_RE = /\b[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}\b/i;

const RANDOM_NUMBER_RE = /\b(?:random|generated)\s*(?::?\s*)\d+\b/i;

export function idempotentResponse(options: IdempotentResponseOptions): Guard {
  return {
    name: 'idempotent-response',
    version: '0.1.0',
    description: 'Detects potentially non-idempotent responses',
    category: 'ai',
    supportedStages: ['output'],
    async check(text: string, _ctx: GuardContext): Promise<GuardResult> {
      const start = performance.now();
      const issues: string[] = [];

      if (TIMESTAMP_PATTERNS.some((p) => p.test(text))) {
        issues.push('timestamp_present');
      }
      if (TEMPORAL_PATTERNS.some((p) => p.test(text))) {
        issues.push('temporal_reference');
      }
      if (UUID_RE.test(text)) {
        issues.push('uuid_present');
      }
      if (RANDOM_NUMBER_RE.test(text)) {
        issues.push('random_number');
      }

      const triggered = issues.length > 0;
      const score = triggered ? Math.min(issues.length / 3, 1.0) : 0;

      return {
        guardName: 'idempotent-response',
        passed: !triggered,
        action: triggered ? options.action : 'allow',
        score,
        latencyMs: Math.round(performance.now() - start),
        details: triggered ? { issues } : undefined,
      };
    },
  };
}
